File: Exavision.Seasense/Client/src/app/materials/capabilities/turret/turret-absolute-position-capability.ts
import { Material } from "../../material";
import { SettingCapability } from "../../settings/setting-capability";
import { Site } from "../../site";
import { CapabilityState } from "../../states/capability-state";
import { TurretAbsolutePositionState } from "../../states/turret/turret-absolute-position-state";
import { Unit } from "../../unit";
import { Capability } from "../capability";
import { CapabilityType } from "../capability-type";

export class TurretAbsolutePositionCapability implements Capability {
  id: string;
  capabilityType: CapabilityType;
  material: Material | undefined;
  pan: number = 0;
  tilt: number = 0;
  constructor(settingCapability: SettingCapability) {
    this.id = settingCapability.id;
    this.capabilityType = CapabilityType.TurretAbsolutePosition;
  }
  setState(valueState: CapabilityState): void {
    if (valueState.$type === 'TurretAbsolutePositionState') {
      let turretAbsolutePositionState: TurretAbsolutePositionState = <TurretAbsolutePositionState>valueState;
      this.pan = turretAbsolutePositionState.pan;
      this.tilt = turretAbsolutePositionState.tilt;     
    }
  }

}

File: Exavision.Seasense.Shared/Capabilities/Turret/TurretAbsolutePositionCapability.cs
using Exavision.Seasense.Shared.Settings;
using Exavision.Seasense.Shared.States;
using Exavision.Seasense.Shared.States.Turret;

namespace Exavision.Seasense.Shared.Capabilities.Turret {
    public class TurretAbsolutePositionCapability : Capability<SettingCapabilityEmpty>, ITurretAbsolutePositionCapability {
        public double Pan { get; protected set; }
        public double Tilt { get; protected set; }

        public TurretAbsolutePositionCapability() {

        }

        public override SettingCapabilityEmpty GetSetting(SettingCapabilityEmpty setting) {

            setting.CapabilityType = CapabilityType.TurretAbsolutePosition;
            return setting;
        }

        public override CapabilityState GetState() {
            TurretAbsolutePositionState state = new TurretAbsolutePositionState() {
                Id = this.Id,
                Pan = this.Pan,
                Tilt = this.Tilt
            };
            return state;
        }
    }
}
